Make transaction layer an attribute

Lojjs opened the pull request:
Based on #10678 , that one should be merged first.

Lojjs committed:
Making transactionLayer an attribute

Lojjs committed:
When removing identical plans, make sure to copy over attributes.

We create copies of identical (same object) plans. We need to make
sure all attributes are also copied over to the new id of the copy.

fickludd commented on community/cypher/cypher-logical-plans-3.4/src/main/scala/org/neo4j/cypher/internal/v3_4/logical/plans/Expand.scala:
```diff
@@ -20,22 +20,16 @@
-                 )(val solved: PlannerQuery with CardinalityEstimation)
-  extends LogicalPlan with LazyLogicalPlan {
+case class Expand(source: LogicalPlan, from: IdName, dir: SemanticDirection, types: Seq[RelTypeName], to: IdName, relName: IdName, mode: ExpansionMode = ExpandAll)(val solved: PlannerQuery with CardinalityEstimation)(implicit idGen: IdGen)
+  extends LogicalPlan(idGen) with LazyLogicalPlan {
```
I would revert this format change, it make this argument list very hard to read. Similar in other places in this file. Auto-format mishap?

fickludd commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/profiler/Profiler.scala:
```diff
@@ -38,9 +38,9 @@ import scala.collection.mutable
-  val rowStats: mutable.Map[LogicalPlanId, ProfilingIterator] = mutable.Map.empty
+  val pageCacheStats: mutable.Map[Id, (Long, Long)] = mutable.Map.empty
+  val dbHitsStats: mutable.Map[Id, ProfilingPipeQueryContext] = mutable.Map.empty
+  val rowStats: mutable.Map[Id, ProfilingIterator] = mutable.Map.empty
```
Seems like these maps could become `Attribute`s later if we want to.

fickludd commented on community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/pipes/Pipe.scala:
```diff
@@ -48,7 +48,7 @@ trait Pipe {
 
   // Used by profiling to identify where to report dbhits and rows
-  def id: LogicalPlanId
+  def id(): Id
```
This could be defined without brackets as `id` should be immutable.

fickludd commented on enterprise/cypher/cypher/src/main/java/org/neo4j/cypher/internal/v3_4/codegen/profiling/ProfilingTracer.java:
```diff
@@ -54,7 +54,7 @@ default double pageCacheHitRatio()
     private final Clock clock;
     private final KernelStatisticProvider statisticProvider;
-    private final Map<LogicalPlanId, Data> data = new HashMap<>();
+    private final Map<Id, Data> data = new HashMap<>();
```
Also candidate for an being an `Attribute`

fickludd commented on enterprise/cypher/physical-planning/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/SlotAllocation.scala:
```diff
@@ -49,8 +50,8 @@ object SlotAllocation {
-  case class PhysicalPlan(slotConfigurations: Map[LogicalPlanId, SlotConfiguration],
-                          argumentSizes: Map[LogicalPlanId, Size])
+  case class PhysicalPlan(slotConfigurations: Map[Id, SlotConfiguration],
+                          argumentSizes: Map[Id, Size])
```
More `Attribute`s, oh yeah.

fickludd commented on community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/TxLayerPlanningIntegrationTest.scala:
```diff
@@ -27,60 +28,68 @@ import org.scalatest.matchers.{MatchResult, Matcher}
 
   test("MATCH and CREATE") {
-    val plan = planFor("MATCH (a) CREATE (b)")._2
+    val tuple = planFor("MATCH (a) CREATE (b)")
```
Is it allowed to type these as `val (_, plan, _, readTxLayerAttribute) = planFor(...)`?

fickludd commented on community/cypher/cypher-planner-3.4/src/main/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/steps/LogicalPlanProducer.scala:
```diff
@@ -36,7 +37,7 @@ import org.neo4j.cypher.internal.v3_4.logical.plans.{DeleteExpression => DeleteE
  * much testing
  */
-case class LogicalPlanProducer(cardinalityModel: CardinalityModel, readTransactionLayer: Int, val idGen : IdGen) extends ListSupport {
+case class LogicalPlanProducer(cardinalityModel: CardinalityModel, readTransactionLayer: Int, readtxLayerAttribute: TransactionLayerAttribute, idGen : IdGen) extends ListSupport {
```
How about changing 
 * `readTransactionLayer` -> `currentTxLayer`
 * `readtxLayerAttribute` -> `txLayers`
 * `TransactionLayerAttribute` -> `TransactionLayer`

sherfert replied:
I agree to shortening and removing the "Attribute" from the names. But I do want to keep "read" in the val name. It makes it very clear that this is the layer an operator reads from, but not the one it writes to (if it is an operator that writes).

fickludd commented on community/cypher/frontend-3.4/src/main/scala/org/neo4j/cypher/internal/frontend/v3_4/PlanningAttributes.scala:
```diff
@@ -0,0 +1,23 @@
+import org.neo4j.cypher.internal.util.v3_4.attribution.Attribute
+
+object PlanningAttributes {
+  class TransactionLayerAttribute extends Attribute[Int]
```
As mentioned in another comment, I suggest we ditch the `...Attribute` suffix from this class. I think it might hurt more than help. Maybe attributes should have plural types to indicate that the instance has the attribute of every node in the tree? E.g. `TransactionLayers` in this case.

fickludd reviewed (approved):
Hey folks, I really like that you are pushing this further!

Suggestions for renamings:
 * `readTransactionLayer` -> `currentTxLayer`
 * `readtxLayerAttribute` -> `txLayers`
 * `TransactionLayerAttribute` -> `TransactionLayers`

I also reviewed #10678 here which is not that smooth. Maybe fix any comments here (if you want to) and merge this instead of the other one. Nice work, will be interesting to see the next steps.

sherfert committed:
Review comments.

sherfert commented:
@fickludd Thanks for the review. I don't see how we can merge this one before #10678. This one is based on having proper IDs on logical plans.

Lojjs committed:
Fix broken test

sherfert committed:
Clean cut between InitialState and LogicalPlanningState

This makes it very clear at which point we go from a state with
no additional attributes to a state that has them. It also allows
to not clutter the BaseState interface and keep PlanningAttributes
in the planner module, where they belong.

fickludd commented:
This PR is a superset of #10678. If this one is merged before #10678 we can just close #10678. When we say it's based on another PR it just means that review for the first part is expected elsewhere, but I messed that up a little =).

fickludd commented on community/cypher/cypher-planner-3.4/src/main/scala/org/neo4j/cypher/internal/compiler/v3_4/phases/LogicalPlanState.scala:
```diff
@@ -61,12 +61,26 @@ case class LogicalPlanState(queryText: String,
                      startPosition = state.startPosition,
                      plannerName = state.plannerName,
-                     readTxLayerAttribute = state.readTxLayerAttribute,
+                     readTxLayerAttribute = new TransactionLayers,
```
Howabout changing this argument to `txLayers` or `transactionLayers` as well?

sherfert committed:
Test fix.

systay commented:
@fickludd Had some interesting comments yesterday, on situations where tx-layers would not be enough to handle all cases. Let's talk some more about that before merging any code around tx-layers.

sherfert commented:
Not happening.
